### ISSUE TITLE
Check bugprone-unchecked-optional-access in Engine clang-tidy.

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#127701]
There are a number of existing violations to work through (https://github.com/flutter/engine/pull/42281), so we'll have to attack this in two stages:
- [ ] Enable the lint, and opt-out files that have warnings to fix
- [ ] Incrementally fix all the files with warnings.

**Comment by chinmaygarde** [comment]
As I said here, we could also opt in all Impeller components by modifying the [`impeller_component`](https://github.com/flutter/engine/pull/42281#issuecomment-1564851994) template.

**Comment by zanderso** [comment]
This check is now enabled in all non-unit test files. Unchecked access is a common pattern in unit tests, and the tests will fail on bad optional accesses, so the lint likely has limited value there. So I'll close this issue as done.

**Comment by flar** [comment]
I'm just running into this now in the new DL geometry class tests where I have methods that return optional values. I would be happy if clang-tidy could be taught to recognize `ASSERT_TRUE(result.has_value());` as protective since it returns if the test fails, but clang-tidy still warns about accesses in the subsequent code. There may be an issue in that `ASSERT_TRUE` is a macro that might not be expanded to return from the function in the environment under which clang-tidy is running? (But I thought it ran under host_debug which should have the macros functional?)

**Comment by flar** [comment]
Meanwhile, I've had good luck with the following macro in my unit test files. I just switch to `ENFORCE_TRUE(result.has_value())` and clang-tidy stops complaining...

```
// ASSERT_TRUE(result.has_value()) should be enough to prevent the test
// code that follows it to run, but clang-tidy doesn't seem to recognize
// that macro expansion as protective. This macro is more proactive
// about ending execution at the place where the statement lives.
#define ENFORCE_TRUE(condition) \
  do {                          \
    if (!(condition)) {         \
      ASSERT_TRUE(condition);   \
      return;                   \
    }                           \
  } while (0)
```

**Comment by zanderso** [comment]
Yeah, due to this problem, we have the clang-tidy check disabled in most(all?) unit test files. If you have a good way forward, we should get that macro added to a common location, so that we can enable the check in more places.
